mbed-drivers/I2C.h

```diff
@@ -26,10 +26,235 @@
+ * and receive events. The I2CTransaction class encapsulates all I2C transaction parameters.
+ * The I2CResourceManager class is a generic interface for implementing I2C resource managers.
+ * This will allow for additional classes of I2C device, for example, a bitbanged I2C master.
+ *
+ * I2C Resource managers are instantiated statically and initialized during global init.
+ * There is one Resource Manager per logical port. Logical ports could consist of:
+ *
+ * * Onchip I2C controllers
+ * * I2C Bridges (SPI->I2C bridge, I2C->I2C bridge, etc.)
+ * * Bit banged I2C
+ * * Bit banged I2C over SPI GPIO expander
+ * * More...
+ *
+ * Currently only onchip I2C controllers are supported.
+ *
+ * ## Constructing I2C transactions
+ * To construct an I2C transaction with more than one segment, use the then member,
+ * or the repeated_start member, with a CriticalSectionLock:
+ *
+ * ```C++
+ * void doneCB(bool dir, I2CBuffer buf, int Event) {
+ *     // Do something
+ * }
+ * I2C i2c0(sda, scl);
+ * I2C i2c1(sda, scl);
+ * void app_start (int, char **) {
+ *     uint8_t cmd[2] = {0xaa, 0x55};
+ *     i2c0.transfer_to(addr).tx(cmd,2).then().rx(4).callback(doneCB, I2C_EVENT_ALL);
+ *     // OR
+ *     CriticalSectionLock lock;
+ *     i2c1.transfer_to(addr).tx(cmd,2).repeated_start();
+ *     i2c1.transfer_to(addr).rx(4).callback(doneCB, I2C_EVENT_ALL);
+ * }
+ * ```
+ */
+
 namespace mbed {
+class I2C;
+/**
+ * The I2CBuffer class is a variant of the Buffer class.
+ * Instead of just storing a buffer pointer and a size, if the buffer is less than
+ * 8 bytes long, it packs the whole buffer into the space occupied by the pointer and
+ * size variable. This is indicated by setting the MSB in size.
+ */
+class I2CBuffer {
+public:
+    /**
+     * Set buffer pointer and length.
+     *
+     * @param[in] b A buffer to duplicate
+     */
+    void set(const Buffer & b);
+    /**
+     * Set the buffer pointer and length
+     *
+     * If the length of the buffer is less than 8 bytes it is stored internally
+     *
+     * @param[in] buf the buffer pointer to duplicate
+     * @param[in] len the length of the buffer to duplicate
+     */
+    void set(void * buf, size_t len);
+    /**
+     * Get a pointer to the buffer.
+     *
+     * If the buffer is the internal storage, return it, otherwise return the reference
+     *
+     * @return the buffer pointer
+     */
+    void * get_buf();
+    /**
+     * Get the length
+     *
+     * @return the length of the buffer
+     */
+    size_t get_len() const;
+protected:
+    union {
+        struct {
+            void * data;
+            size_t len;
+        } _ref;
+        struct {
+            uint8_t data[sizeof(_ref)-1];
+            unsigned small:1;
+            unsigned len:7;
+        } _packed;
+    };
+};
+class I2CTransaction;
+
+namespace detail {
+/**
+ */
+enum class I2CDirection {Transmit, Receive};
+
+class I2CSegment : public I2CBuffer {
```

Comment:
This relationship looks more like composition than inheritance. Not a big problem though.